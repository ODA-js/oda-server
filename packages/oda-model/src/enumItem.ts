import {
  ModelBase,
  ModelBaseInput,
  ModelBaseInternal,
  IModelBase,
  ModelBaseOutput,
  ModelBaseMetaInfo,
} from './modelbase';
import { merge } from 'lodash';
import { MetaModelType, Nullable, assignValue } from './types';
import { Internal } from './element';

export interface IEnumItem
  extends IModelBase<EnumItemMetaInfo, EnumItemInput, EnumItemOutput> {
  readonly value: string;
}

export interface EnumItemMetaInfo extends ModelBaseMetaInfo {}

export interface EnumItemInternal extends ModelBaseInternal {
  value: string;
}

export interface EnumItemInput extends ModelBaseInput<EnumItemMetaInfo> {
  value?: string;
}
export interface EnumItemOutput extends ModelBaseOutput<EnumItemMetaInfo> {
  value?: string;
}

export const enumItemDefaultMetaInfo = {};
export const enumItemDefaultInput = { metadata: enumItemDefaultMetaInfo };

export class EnumItem
  extends ModelBase<
    EnumItemMetaInfo,
    EnumItemInput,
    EnumItemInternal,
    EnumItemOutput
  >
  implements IEnumItem {
  public get modelType(): MetaModelType {
    return 'enum-item';
  }

  constructor(init: EnumItemInput) {
    super(merge({}, enumItemDefaultInput, init));
  }

  get value(): string {
    return this[Internal].value || this[Internal].name;
  }

  updateWith(input: Nullable<EnumItemInput>) {
    super.updateWith(input);
    assignValue<EnumItemInternal, EnumItemInput, EnumItemInput['value']>({
      src: this[Internal],
      input,
      field: 'value',
    });
  }

  toObject(): EnumItemOutput {
    return merge({}, super.toObject(), {
      value: this[Internal].value,
    } as Partial<EnumItemOutput>);
  }
  mergeWith(payload: Nullable<EnumItemInput>) {
    super.mergeWith(payload);
  }
}
