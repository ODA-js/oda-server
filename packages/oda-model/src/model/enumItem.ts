import {
  MetaModelType,
  EnumItemMeta,
  EnumItemInput,
  EnumItemStorage,
  IEnumItem,
} from './interfaces';
import { ModelBase } from './modelbase';

export class EnumItem
  extends ModelBase<EnumItemMeta, EnumItemInput, EnumItemStorage>
  implements IEnumItem {
  public modelType: MetaModelType = 'enum-item';
  constructor(inp: EnumItemInput) {
    super(inp);
  }
  get value(): string {
    return this.$obj.value;
  }
  updateWith(inp: EnumItemInput) {
    super.updateWith(inp);
    this.$obj.value = inp.value || '';
  }
  toObject(): EnumItemInput {
    return {
      ...super.toObject(),
      value: this.value,
    };
  }
}
