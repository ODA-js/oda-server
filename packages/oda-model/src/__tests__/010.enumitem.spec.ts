import 'jest';
import { EnumItem, enumItemDefaultMetaInfo } from '../enumItem';

describe('enumItem', () => {
  it('default', () => {
    const res = new EnumItem({ name: 'ITEM', value: 'I_T_E_M' });
    expect(res.metadata).toMatchObject(enumItemDefaultMetaInfo);
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
});
