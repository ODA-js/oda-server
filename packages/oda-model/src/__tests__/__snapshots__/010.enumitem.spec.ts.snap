// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enumItem default: toObject 1`] = `
Object {
  "description": undefined,
  "metadata": Object {},
  "name": "ITEM",
  "title": undefined,
  "value": "I_T_E_M",
}
`;
